forage-android/src/main/java/com/joinforage/forage/android/collect/BTPinCollector.kt

```diff
@@ -0,0 +1,119 @@
+package com.joinforage.forage.android.collect
+
+import com.basistheory.android.service.BasisTheoryElements
+import com.basistheory.android.service.ProxyRequest
+import com.joinforage.forage.android.BuildConfig
+import com.joinforage.forage.android.network.model.ForageApiError
+import com.joinforage.forage.android.network.model.ForageApiResponse
+import com.joinforage.forage.android.network.model.ForageError
+import com.joinforage.forage.android.ui.ForagePINEditText
+import kotlinx.coroutines.runBlocking
+import org.json.JSONException
+import kotlin.coroutines.suspendCoroutine
+
+internal class BTPinCollector(
```

Comment:
We need to change this guy a little bit.

On VGS implementation, they used a really old approach to send the requests. The `suspendCoroutine` function was necessary to transform the callbacks approach into suspending function.

Also, BT documentation didn't help much here; usually, `runBlocking` should be used only for tests.

One approach we can use here is to get rid of the `suspendCoroutine` function since `bt.proxy.post` is already a suspending function:

```
    override suspend fun collectPinForBalanceCheck(
        paymentMethodRef: String,
        cardToken: String,
        encryptionKey: String
    ): ForageApiResponse<String> {
        val bt = buildBt()

        val proxyRequest: ProxyRequest = ProxyRequest().apply {
            headers = mapOf(
                "X-KEY" to encryptionKey,
                "Merchant-Account" to merchantAccount,
                "BT-PROXY-KEY" to PROXY_ID,
                "Content-Type" to "application/json"
            )
            body = object {
                val pin = pinForageEditText.getTextElement()
                val card_number_token = cardToken
            }
            path = balancePath(paymentMethodRef)
        }

        pinForageEditText.getTextElement().setText("")

        return runCatching {
            bt.proxy.post(proxyRequest)
        }.fold(
            onSuccess = {
                val forageApiError = ForageApiError.ForageApiErrorMapper.from(it.toString())
                val error = forageApiError.errors[0]

                // our logic to check errors in case the request succeeded
                ForageApiResponse.Failure(listOf(ForageError(400, error.code, error.message)))

            },
            onFailure = {
                // our logic to check errors in case the request failed
                ForageApiResponse.Failure(emptyList())
            }
        )
    }
```

Reply:
we can follow this on `collectPinForBalanceCheck` & `collectPinForCapturePayment`

Reply:
Thank you for the clarification! `suspendCoroutine` was removed here: [7db5c7e](https://github.com/teamforage/forage-android-sdk/pull/38/commits/7db5c7e1321e735be25afc8b46fab634ed71d8ce)